Add `cache_subnet_group_name` to elasticache module

quodlibetor opened the pull request:
According to the [AWS docs on elasticache and the VPC](http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/CacheSubnetGroups.html) subnet groups are required inside VPCs.

Commit:
Add `cache_subnet_group` to elasticache module

According to the [docs] cache subnet groups are required inside a VPC.

[docs]: http://docs.aws.amazon.com/AmazonElastiCache/latest/UserGuide/CacheSubnetGroups.html

rrana commented:
Any progress on this?

craigmarvelley commented:
I'd really appreciate this being merged in once the conflicts are resolved. Happy to do that and resubmit if @quodlibetor isn't able to?

quodlibetor commented:
I'm extremely willing to do any work to help get this merged :)

quodlibetor commented:
Also, FWIW the way that I'm working around not having this merged upstream is just having a copy of the fixed version in my `library/cloud/` dir, which the default `ansible.cfg` has as a higher priority than the global one.

bcoca commented:
@quodlibetor can you rebase, this PR currently does not merge cleanly

quodlibetor commented:
Rebased, no conflicts when rebasing. :8ball:

quodlibetor commented:
Any other problems?

InfiniteLee commented:
Bump for merging!

rouge8 commented:
I don't think this actually works, because `cache_security_groups` will always be set to `['default']`, which is invalid within a VPC.

rouge8 commented:
Nevermind, you can set `cache_security_groups: []`.

barraponto commented:
and is there a way to create a cache subnet group using ansible?

juliedavila commented:
@quodlibetor this looks good, can you change the parameter to `cache_subnet_group` instead of being `cache_subnet_group_name`
@rouge8 Hmm, seems hacky, I'll refactor this later so it can detect if it's going to go into a vpc and trigger a `[]` default.
@barraponto not atm, but will look into adding that into the module elasticache after the PR is merged.

quodlibetor commented:
Updated to use `cache_subnet_group` and rebased.

Commit:
Make cache_subnet_group default to empty security groups

Since they can't be specified together, it seems like setting the
default to no security groups is the right option.

quodlibetor commented:
I've added a change that defaults `cache_security_groups` to an empty list if nothing was provided.

I used a sentinel value to set the default since it's plausible that people are explicitly setting the list to `default`, in which case they would get some unexpected behavior if we just overwrote it. If the argument spec is ever serialized to json this would cause a ValueError traceback, I believe. I don't know how likely that is. 

I'm willing to change it to use the less-robust option if that's what you'd like.

juliedavila commented:
@quodlibetor 

Unfortunately that doesn't seem to work. With the sentinel value it gives the following error:

``` shell
GATHERING FACTS ***************************************************************
ok: [localhost]

TASK: [Elasticache] ***********************************************************
failed: [localhost] => {"failed": true}
msg: argument cache_security_groups is of invalid type: <type 'object'>, required: list

FATAL: all hosts have already failed -- aborting

PLAY RECAP ********************************************************************
           to retry, use: --limit @/Users/jdavila/aws.retry

localhost                  : ok=1    changed=0    unreachable=0    failed=1
```

And for the sake of trying I attempted to cast default as a list and it returned

```
PLAY [Test AWS Stuffs] ********************************************************

GATHERING FACTS ***************************************************************
ok: [localhost]

TASK: [Elasticache] ***********************************************************
failed: [localhost] => {"failed": true, "parsed": false}
Traceback (most recent call last):
  File "/Users/jdavila/.ansible/tmp/ansible-tmp-1427391056.94-38111842336176/elasticache", line 2365, in <module>
    main()
  File "/Users/jdavila/.ansible/tmp/ansible-tmp-1427391056.94-38111842336176/elasticache", line 498, in main
    cache_security_groups={'required': False, 'default': list(default),
TypeError: 'object' object is not iterable


FATAL: all hosts have already failed -- aborting

PLAY RECAP ********************************************************************
           to retry, use: --limit @/Users/jdavila/aws.retry

localhost                  : ok=1    changed=0    unreachable=0    failed=1
```

The less robust option might be the answer here, though I'm open to a different work around.

quodlibetor commented:
Your specific issue comes from the fact that `list()` actually creates a list by iterating over the contents of its argument, using `[default]` just wraps `default` in a list, which does work and is in the lastest version.

The newest version was tested with the following playbook:

```
---
- hosts: localhost
  tasks:
    - local_action:
        module: elasticache
        name: test-bwm
        state: present
        cache_subnet_group: doesnotexist
        num_nodes: 1
        region: us-east-1
```

And resulted in:

```
$ ansible-playbook -i ec2.py test.yml

PLAY [localhost] **************************************************************

GATHERING FACTS ***************************************************************
ok: [localhost]

TASK: [elasticache] ***********************************************************
failed: [localhost -> 127.0.0.1] => {"failed": true}
msg: Cache Subnet Group doesnotexist does not exist.
```

Which is a correct error that comes from the call boto, so it passes all of the ansible verification now.

quodlibetor commented:
Rebased.

juliedavila commented:
Tested. Good to go. Thanks.

quodlibetor commented:
:sunrise:
